### PR TITLE
Add Box::leak<'a>(Box<T>) -> &'a mut T where T: 'a

### PR DESCRIPTION
**Pull request by Centril**
Adds:

```rust
impl<T: ?Sized> Box<T> {
    pub fn leak<'a>(b: Box<T>) -> &'a mut T where T: 'a {
        unsafe { &mut *Box::into_raw(b) }
    }
}
```

which is useful for when you just want to put some stuff on the heap and then have a reference to it for the remainder of the program.

r? @sfackler 
cc @durka

### CODE REVIEW

**Comment by durka** [comment]
N.B. this exists in the ecosystem: https://crates.io/crates/leak

Is `&'static mut T` useful? It's unsafe to do anything with that, right? Maybe `&'static T` is better.

**Comment by mattico** [comment]
See also: https://github.com/rust-lang/rfcs/pull/1233

**Comment by oli-obk** [comment]
It's not unsafe. There's a single mutable reference, the lifetime is irrelevant

**Comment by Centril** [comment]
Hmm, so @alexcrichton said then in rust-lang/rfcs#1233:
> While this is indeed a pattern that could be added to the standard library, it's currently not greatly motivated in terms of use cases that require this sort of functionality to be in the standard library itself.

Since then the [`leak`](https://crates.io/crates/leak) crate has been used: `12,959` times, but has one immediate dependent crate.

**Comment by cramertj** [comment]
> Since then the leak crate has been used: 12,959 times, but has one immediate dependent crate.

And that crate is transitively depended on by `skia` and `servo-skia`-- I'd be that's where many of the downloads are coming from.

**Comment by Centril** [comment]
@cramertj So I can't judge if this means that the feature is used very little or not... =)

**Comment by sfackler** [comment]
I think this is a nice addition. `&'static mut T` seems like the right type, but `'static` mutability is a bit subtle. @alexcrichton do you know if this'll run into any safety issues?

@rfcbot fcp merge

**Comment by alexcrichton** [comment]
AFAIK there's no soundness holes from this, but we can always double check during stabilization!

**Comment by xfix** [comment]
I wonder if there is no risk of BC break in event this function would be replaced with this.

```rust
fn leak<'a>(b: Box<T>) -> &'a mut T
where
    T: 'a,
{
    unsafe { &mut *Box::into_raw(b) }
}
```

**Comment by retep998** [comment]
I have two points that people may be concerned about and it would be nice to have clear answers about them:

* What if `T` is not `'static`? Will the returned `&mut 'static T` still respect the lifetime of `T` itself?
* What happens when `Box` gains the ability to be parametric over the allocator and that allocator has a lifetime so when the allocator dies, it takes all its allocations with it including any "leaked" allocations.

**Comment by Centril** [comment]
@xfix So; talked it over on #rust-lang and it seems strictly more powerful - so I'll change it to your proposal `fn leak<'a, T: 'a>(b: Box<T>) -> &'a mut T` instead.

@retep998 

> What if T is not 'static? Will the returned &mut 'static T still respect the lifetime of T itself?

Given [this program](https://play.rust-lang.org/?gist=3a7dbd85329fe91c19e245cf6678f241&version=stable):
```rust
fn leak<T>(b: Box<T>) -> &'static mut T {
    unsafe { &mut *Box::into_raw(b) }
}

fn main() {
    let static_ref: &'static mut &mut usize = {
        let mut local_variable = 41;
        let boxed: Box<&mut usize> = Box::new(&mut local_variable);
        leak(boxed)
    };
    **static_ref += 1;
    assert_eq!(**static_ref, 42);
}
```
we, rightly, get:
```
error[E0597]: `local_variable` does not live long enough
```
So a bound `T: 'static` is magically there.

> What happens when Box gains the ability to be parametric over the allocator...

Well, then `leak` does not exist for that `Box<T, Alloc>` yet, but can be defined later by `Alloc: 'a`.

**Comment by Centril** [comment]
Updated to `fn leak<'a, T: 'a>(b: Box<T>) -> &'a mut T` after OKing with @sfackler.

**Comment by carols10cents** [comment]
ping @BurntSushi for your ticky box [here!](https://github.com/rust-lang/rust/pull/45881#issuecomment-343041780)

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -364,6 +364,59 @@ impl<T: ?Sized> Box<T> {
...
+    ///
+    /// Note: this is an associated function, which means that you have
+    /// to call it as `Box::leak(b)` instead of `b.leak()`. This
+    /// is so that there is no conflict with a method on the inner type.
+    ///
+    /// [`Box::from_raw`]: struct.Box.html#method.from_raw
+    ///
+    /// # Examples
+    ///
+    /// Simple usage:
```
Comment (xfix) [inline src/liballoc/boxed.rs]: Is a documentation line simply saying "Simple usage" before an example needed?

Comment (Centril) [reply src/liballoc/boxed.rs]: I used it to differentiate between "simple usage" and "unsized data" - I think it makes it a bit more readable - but I can always remove it.

**Comment by BurntSushi** [comment]
If `'static` is there implicitly, does it make sense to make it explicit with `T: 'static` in the function signature?

**Comment by Centril** [comment]
@BurntSushi I generalized it to any lifetime `'a` and made it explicit after popular demand, so yes =)

**Comment by BurntSushi** [comment]
@Centril Ah great!

**Comment by alexcrichton** [comment]
@Centril mind updating the issue number reference in the unstable tag? After that this should be good to go!

**Comment by Centril** [comment]
@alexcrichton Done, issue is #46179 .

**Pull request merged**